CIP-0144/README.md

```diff
@@ -0,0 +1,356 @@
+
+We will use an identifier in the anchor to refer to the schema where each type is defined. For example, if we want to reference the `Transaction` type, as defined in CIP-116 we will use the following schema reference `{ "$ref": "#/cip-116/Transaction" }`.
+
+If an operation does not require any arguments as part of it's request, or does not return a meaningful response, we will represent this as an `{}` on the respective operation field. The arguments to an operation will always be represented as an object. If a field is not marked as `required`, then that argument is to be considered optional.
+
+For each operation we will provide some details on how the implementation should behave. Note that some of these are taken verbatim from CIP-30.
+
+### Connection API
+
+The role of the connection API is to provide generic information about the wallet and to allow the user to opt into the functionalities (or extensions) that they want the wallet to provide.
+
+##### Enable
+
+```
+{
+  "operation": "enable",
+  "request": { "$ref": "#/appendix/Extensions" },
+  "response": {},
+  "errors": [
+    { "$ref": "#/appendix/APIError" }
+  ]
+}
+```
+
+This is the entrypoint to start communication with the user's wallet. The wallet should request the user's permission to connect the web page to the user's wallet, and if permission has been granted, the APIs will be made available for the dApp to use. The wallet can choose to maintain a whitelist to not necessarily ask the user's permission every time access is requested, but this behavior is up to the wallet and should be transparent to web pages using this API. If a wallet is already connected this function should not request access a second time.
+
+Through the `extension` field, dApps can request a list of what functionality they expect as a list of CIP numbers (and optional versions) capturing those extensions. This is used as an extensibility mechanism to document what functionalities can be provided by the wallet interface. We will see later in this document examples of what such extensions might be and which functionalities they will enable. New functionalities can be introduced via additional CIPs and may be all or partially supported by wallets.
+
+When requesting the functionalities of a cip extension, dApps can optionally specify a version number for it. Every cip extension defined in the future, must also define a versioning scheme following SemVer. When the version argument is not specified, wallets should take that as the greater version they implement. In this context - and for the rest of this discussion - we will assume versions are ordered with the canonical ordering of SemVer. If the dApp requests a specific version of an extension, wallets can only accept the request if a version of the extension they implement has:
+
+- the same major value as the one requested by the dApp AND
```

Comment:
do we need to include SemVer versioning for extensions?
I feel like this maybe adding unneeded complexity

Existing CIP30 extensions, aren't versioned, instead to replace one you must provide another CIP
I think this is mostly fine

Reply:
In prior discussions for both CIPs and CIP-0030 extensions for last 4 years or so, in my recollection all parties agreed each time that an integer "serial number" was both adequate and future-proof as a version ID _except_ for those who argued that "semantic versioning is _always_ preferred" as a general principle.

Reply:
I think we should have the option to update an API through versions. It enables a more 'agile' approach to defining and implementing CIPs